ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py

```diff
@@ -0,0 +1,283 @@
+#!/usr/bin/python
+
+# Copyright 2020 Hewlett Packard Enterprise Development LP
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+# file except in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
+# OF ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+# author alok ranjan (alok.ranjan2@hpe.com)
+
+# this file will ultimately sit in "/usr/lib/python3.6/site-packages/ansible/module_ in production
+import datetime
+import uuid
+
+
+def is_null_or_empty(name):
+    if type(name) is bool:
+        return False
+    if name is None or not name or name == "":
+        return True
+    return False
+
+
+def get_unique_string(baseName):
+    unique_string = baseName + datetime.datetime.now().strftime(
+        "-%d-%m-%Y") + '-' + str(uuid.uuid1().time)
+    return unique_string
+
+
+# remove arguments from kwargs which are by default none or empty
+def remove_null_args(**kwargs):
+    tosearch = kwargs.copy()
+    for key, value in tosearch.items():
+        if type(value) is not bool:
+            if is_null_or_empty(value):
+                kwargs.pop(key)
+    return kwargs
+
+
+def is_dict_same(server_dict, dict_to_check):
+    if server_dict is None and dict_to_check is None:
+        return True
+    for key_to_check in dict_to_check.keys():
+        # there can be two possibilities.
+        # 1. key is not present. hence return false.
+        # 2. key is present, but value is not same. return false for this too
+        if key_to_check in server_dict.keys():
+            if dict_to_check[key_to_check] == server_dict[key_to_check]:
+                continue
+            else:
+                return False
+        else:
+            # key not present
+            return False
```

Comment:
ok so the intention of the function is to check if the key ,value pair of "dict_to_check" is present in server_dict. 
for ex :
dict_to_check = { "key1" , "value1"}
server_dict = {"key1, "value1", "key2" , "value2"}

in the above case the call to 
```
if (0 if dict1 is None else len(dict1)) != (0 if dict2 is None else len(dict2)):
          return False
```
will return false. but this is not what we want. i want to check if the key value present in dict_to_check is there in server_dict or not.

I guess the name "is_dict_same()" is not correct and is creating confusion.
i have now renamed it to "is_dict_key_value_present_in_server_dict" . hope that will make the intention clear.

1.agreed, if dict_to_check is empty we were returning true. i have put a check now
```

if len(dict_to_check) == 0:
        return False

```

Reply:
also, 
i have taken this piece of code as you have pointed.. looks much neat and clean 
```
for key, value in dict1.items():
         if key not in dict2 or value != dict2[key]:
```

Reply:
okay, so this function is intending to check if subset dict items exists within large dictionary.
I think in that case, we can use the dictionary comparison operator `<=` as follows:

```
def is_subset_dict(dict1, dict2):
    return dict1.items() <= dict2.items()

d1 = {'k1':'v1', 'k2':'v2', 'k3':'v3'}
d2 = {'k1':'v1', 'k2':'v2', 'k3':'v3', 'k4':'v4'}
d3 = {'k3':'v3'}
d4 = {'k5': 'v5'}
d5 = {}

print("********************")
print(is_subset_dict(d1, d2)) # True
print(is_subset_dict(d2, d3)) # False
print(is_subset_dict(d3, d1)) # True
print(is_subset_dict(d3, d2)) # True
print(is_subset_dict(d4, d1)) # False 
print(is_subset_dict(d5, d1)) # True
```

Reply:
agreed that subset dict() will also give the desired result. but I personally like the below logic .

```
    for key, value in dict_to_check.items():        
        if key not in server_dict or value != server_dict[key]:
            return False
    return True
```

reasons:
1. the subset logic() is valid for 3.6 but if the same code is copied and ported on 2.6 or lower it will break. https://stackoverflow.com/a/41579450/2345891 .
I know that we are using 3.6 and above and hence we are safe here but my take is if there is an alternate solution which works on every version and without any performance degradation then we should go for that.
2. i found the above code is more easy to understand and is more readable. it's not that 
is_subset_dict() is hard to understand but then that's my personal view.the logic is more explicit and one can easily figure out the intention on what is being checked.

Do you see any major concern if we follow the logic i have pasted.
By the way i have already modified and refactored it as per your suggestion and it looks much neat and clean now

```

def is_dict_key_value_present_in_server_dict(server_dict, dict_to_check):
    if server_dict is None and dict_to_check is None:
        return True
    if len(dict_to_check) == 0:
        return False

    for key, value in dict_to_check.items():
        if key not in server_dict or value != server_dict[key]:
            return False
    return True
```

Reply:
We will be supporting ansible-modules for 3.6 or later version only. So, we can ignore the older python versions.
I think we should better use the built-in comparisons as there are better optimized for performance and well-tested.
In the above case, we can directly call `dict1.items() <= dict2.items()` in the caller itself instead of another function`is_subset_dict() or is_dict_key_value_present_in_server_dict()` unless we need to add other validations if any (in case not considered by built-in comparison).